google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/SpannerOptions.java

```diff
@@ -179,6 +206,99 @@ public Builder setSessionLabels(Map<String, String> sessionLabels) {
+     * InstanceAdminClient}. Instead of calling this method, you should set specific {@link
+     * RetrySettings} for each of the underlying gRPC clients by calling respectively {@link
+     * #spannerStubSettingsBuilder()}, {@link #databaseAdminStubSettingsBuilder()} or {@link
+     * #instanceAdminStubSettingsBuilder()}.
+     */
+    @Override
+    public Builder setRetrySettings(RetrySettings retrySettings) {
+      throw new UnsupportedOperationException(
+          "SpannerOptions does not support setting global retry settings. "
+              + "Call spannerStubSettingsBuilder().<method-name>Settings().setRetrySettings(RetrySettings) instead.");
+    }
+
+    /**
+     * Returns the {@link SpannerStubSettings.Builder} that will be used to build the {@link
+     * SpannerRpc}. Use this to set custom {@link RetrySettings} for individual gRPC methods.
+     *
+     * <p>The library will automatically use sensible defaults if no custom settings are set. The
+     * defaults are the same as the defaults that are used by {@link SpannerSettings}. Retries are
+     * configured for idempotent methods but not for non-idempotent methods.
+     *
+     * <p>You can set the same {@link RetrySettings} for all unary methods by calling this:
+     *
+     * <pre><code>
+     * builder
+     *     .spannerStubSettingsBuilder()
+     *     .applyToAllUnaryMethods(
+     *         new ApiFunction&lt;UnaryCallSettings.Builder&lt;?, ?&gt;, Void&gt;() {
+     *           public Void apply(Builder&lt;?, ?&gt; input) {
+     *             input.setRetrySettings(retrySettings);
+     *             return null;
+     *           }
+     *         });
+     * </code></pre>
+     */
+    public SpannerStubSettings.Builder spannerStubSettingsBuilder() {
+      return spannerStubSettingsBuilder;
+    }
+
+    /**
+     * Returns the {@link InstanceAdminStubSettings.Builder} that will be used to build the {@link
+     * SpannerRpc}. Use this to set custom {@link RetrySettings} for individual gRPC methods.
+     *
+     * <p>The library will automatically use sensible defaults if no custom settings are set. The
+     * defaults are the same as the defaults that are used by {@link InstanceAdminSettings}. Retries
+     * are configured for idempotent methods but not for non-idempotent methods.
+     *
+     * <p>You can set the same {@link RetrySettings} for all unary methods by calling this:
+     *
+     * <pre><code>
+     * builder
+     *     .instanceAdminStubSettingsBuilder()
+     *     .applyToAllUnaryMethods(
+     *         new ApiFunction&lt;UnaryCallSettings.Builder&lt;?, ?&gt;, Void&gt;() {
+     *           public Void apply(Builder&lt;?, ?&gt; input) {
+     *             input.setRetrySettings(retrySettings);
+     *             return null;
+     *           }
+     *         });
+     * </code></pre>
+     */
+    public InstanceAdminStubSettings.Builder instanceAdminStubSettingsBuilder() {
```

Comment:
Taking a step back, I find it confusing that there are getters on a builder (IMO). What do you think about adding setting APIs as part of the SpannerOptions Builder, and just setting defaults if they are not set by the user? CC @sduskis for your thoughts on how the API should look here.

Reply:
I based this idea on how the (generated) `SpannerSettings` for `SpannerClient` does this with 'recursive builders': https://github.com/googleapis/google-cloud-java/blob/d931fac64e331bbcffea7ff9dbd271e31bac5e05/google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/v1/SpannerSettings.java#L78

We will need some sort of defaults, as gax requires it (even if the default is 'do not retry' or 'retry without ever giving up').

Reply:
Ah I see - in that case I would either add 'get' or 'create' to these to be a little clearer

Reply:
Changed to get...SettingsBuilder(). 'create' would not be appropriate here, as the settings builders have already been created with their default values, and these methods get these builders.